### PR TITLE
feat: Adds support for selecting emojis using the keyboard

### PR DESCRIPTION
**Pull request by iamsivin**
# Pull Request Template

## Description

1. This PR will add support for select emoji from keyboard by adding `:{search emoji name}`. 
2. Moved insert special content (Mention, canned response, variable and emoji selection) methods to helper. 
3. And made some changes in the **Prosemirror** repo to support min characters in `triggerCharacters` method, here is the PR https://github.com/chatwoot/prosemirror-schema/pull/30

Fixes https://linear.app/chatwoot/issue/CW-3549/add-support-for-selecting-emojis-using-the-keyboard

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

**Screen recording**

https://github.com/user-attachments/assets/65d9fd34-44a0-4bd1-9da3-ab652ddeb2b3

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/javascript/dashboard/helper/editorHelper.js [hunk 1]
```diff
@@ -281,3 +283,109 @@ export function setURLWithQueryAndSize(selectedImageNode, size, editorView) {
...
+/**
+ * Creates a content node based on the type (mention, canned response, variable, or emoji)
+ * @param {Object} editorView - The editor view instance
+ * @param {string} type - The type of content (mention, cannedResponse, variable, or emoji)
+ * @param {string|Object} content - The content to create the node from
+ * @param {Object} range - The range object containing the from and to positions
+ * @param {Object} variables - The variables to replace in the canned response (optional)
+ * @returns {Object} - The created content node and the updated from and to positions
+ */
+export const getContentNode = (editorView, type, content, range, variables) => {
```
Comment (muhsin-k) [inline app/javascript/dashboard/helper/editorHelper.js:370]: I believe there's some repetition in the node creation functions that we can refactor. Also the `getContentNode` function can be simplified using a more functional approach

```
const createNode = (editorView, nodeType, content) => {
  const { state } = editorView;
  switch (nodeType) {
    case 'mention':
      return state.schema.nodes.mention.create({
        userId: content.id,
        userFullName: content.name,
      });
    case 'cannedResponse':
      return new MessageMarkdownTransformer(messageSchema).parse(content);
    case 'variable':
      return state.schema.text(`{{${content}}}`);
    case 'emoji':
      return state.schema.text(content);
    default:
      return null;
  }
};

const nodeCreators = {
  mention: (editorView, content, from, to) => ({ node: createNode(editorView, 'mention', content), from, to }),
  cannedResponse: (editorView, content, from, to, variables) => {
    const updatedMessage = replaceVariablesInMessage({ message: content, variables });
    const node = createNode(editorView, 'cannedResponse', updatedMessage);
    return { node, from: node.textContent === updatedMessage ? from : from - 1, to };
  },
  variable: (editorView, content, from, to) => ({ node: createNode(editorView, 'variable', content), from, to }),
  emoji: (editorView, content, from, to) => ({ node: createNode(editorView, 'emoji', content), from, to }),
};

export const getContentNode = (editorView, type, content, { from, to }, variables) => {
  const creator = nodeCreators[type];
  return creator ? creator(editorView, content, from, to, variables) : { node: null, from, to };
};

```

Note: Please test it properly updating the code.

**Comment by muhsin-k** [comment]
Good work ❤️ . I have 2 minor suggestions,

1/ Can we improve the search experience. I searched `sad` and got different results.

CleanShot 2024-09-03 at 12 53 24@2x

2/ The emoji selector is not closing when you navigate to different conversation.

https://github.com/user-attachments/assets/498807fd-716c-4517-8897-bf31fa9461e5

File: app/javascript/dashboard/helper/editorHelper.js (hunk 1 above)
Comment (iamsivin) [reply app/javascript/dashboard/helper/editorHelper.js]: @muhsin-k Done https://github.com/chatwoot/chatwoot/pull/10055/commits/109a1400a6362acbd6437b3c0e3b9abffb339e49

**Comment by iamsivin** [comment]
> Good work ❤️ . I have 2 minor suggestions,
> 
> 1/ Can we improve the search experience. I searched `sad` and got different results.
> 
@muhsin-k I tried searching `sad`. It's showing for me
image

> 
> 2/ The emoji selector is not closing when you navigate to different conversation.
> 
Fixed this issue for both emoji and variable selector https://github.com/chatwoot/chatwoot/pull/10055/files#diff-f1a495eddfdad6afbc89198872baa057a05195ea973422e0257545866b8bb5baR303 .

**Review by muhsin-k — APPROVED** [review]

**Pull request merged**